core/src/main/java/org/elasticsearch/plugins/Plugin.java

```diff
@@ -123,6 +125,16 @@ public void onIndexModule(IndexModule indexModule) {}
     public List<String> getSettingsFilter() { return Collections.emptyList(); }
 
     /**
+     * Returns a function that upgrades metadata on startup.
+     * Plugins should return the input metadata if no upgrade is required
+     * and throw {@link IllegalStateException} to stop a node from starting
+     * when unsupported metadata is found.
```

Comment:
I suggest more official Javadoc:

```java
/**
 * Provides a function to modify the metadata on startup.
 * <p>
 * Plugins should return the input metadata via {@link Function#identity()} if no upgrade is required.
 * @return Never {@code null}. The same or upgraded {@code MetaData}.
 * @throws IllegalStateException if the node needs to be stopped because {@code MetaData} is unsupported
 */
```